Mathlib/Tactic/CC/Datatypes.lean

```diff
@@ -0,0 +1,653 @@
+      we store `tgt` at `target`, and `H` at `proof`. Both fields are none if `e == root` -/
+  target : Option Expr := none
+  /-- When `e` was added to this equivalence class because of an equality `(H : e = tgt)`, then
+      we store `tgt` at `target`, and `H` at `proof`. Both fields are none if `e == root` -/
+  proof : Option EntryExpr := none
+  /-- Variable in the AC theory. -/
+  acVar : Option Expr := none
+  /-- proof has been flipped -/
+  flipped : Bool
+  /-- `true` if the node should be viewed as an abstract value -/
+  interpreted : Bool
+  /-- `true` if head symbol is a constructor -/
+  constructor : Bool
+  /-- `true` if equivalence class contains lambda expressions -/
+  hasLambdas : Bool
+  /-- `heqProofs == true` iff some proofs in the equivalence class are based on heterogeneous
+      equality. We represent equality and heterogeneous equality in a single equivalence class. -/
+  heqProofs : Bool
+  /-- If `fo == true`, then the expression associated with this entry is an application, and we are
+      using first-order approximation to encode it. That is, we ignore its partial applications. -/
+  fo : Bool
+  /-- number of elements in the equivalence class, it is meaningless if `e != root` -/
+  size : Nat
+  /-- The field `mt` is used to implement the mod-time optimization introduce by the Simplify
+      theorem prover. The basic idea is to introduce a counter gmt that records the number of
+      heuristic instantiation that have occurred in the current branch. It is incremented after each
+      round of heuristic instantiation. The field `mt` records the last time any proper descendant
+      of of thie entry was involved in a merge. -/
+  mt : Nat
+  deriving Inhabited
+
+/-- Stores equivalence class data associated with an expression `e`. -/
+abbrev Entries := RBExprMap Entry
+
+/-- Equivalence class data associated with an expression `e` used by AC congruence closure
+modules. -/
+structure ACEntry where
+  /-- Natural number associated to an expression. -/
+  idx : Nat
+  RLHSOccs : RBACAppsSet := ∅
+  RRHSOccs : RBACAppsSet := ∅
+  deriving Inhabited
+
+def ACEntry.ROccs (ent : ACEntry) : (inLHS : Bool) → RBACAppsSet
+  | true => ent.RLHSOccs
+  | false => ent.RRHSOccs
+
+structure ParentOcc where
+  expr : Expr
+  /-- If `symmTable` is true, then we should use the `symmCongruences`, otherwise `congruences`.
+      Remark: this information is redundant, it can be inferred from `expr`. We use store it for
+      performance reasons. -/
+  symmTable : Bool
+
+/-- Red-black sets of `ParentOcc`s. -/
+abbrev ParentOccSet := Std.RBSet ParentOcc (Ordering.byKey ParentOcc.expr compare)
+
+abbrev Parents := RBExprMap ParentOccSet
+
+inductive CongruencesKey
+  | fo (fn : Expr) (args : Array Expr) : CongruencesKey
+  | ho (fn : Expr) (arg : Expr) : CongruencesKey
+  deriving BEq, Hashable
+
+abbrev Congruences := Std.HashMap CongruencesKey (List Expr)
```

Comment:
```suggestion
/-- Maps each expression (via `mkCongruenceKey`) to expressions it might be congruent to. -/ 
abbrev Congruences := Std.HashMap CongruencesKey (List Expr)
```